src/ILCompiler.Compiler/src/IL/ILImporter.Scanner.cs

```diff
@@ -111,8 +111,13 @@ public DependencyList Import()
                 if (!_isFallbackBodyCompilation &&
                     (_canonMethod.Signature.IsStatic || _canonMethod.IsConstructor || owningType.IsValueType))
                 {
-                    // For beforefieldinit, we can wait for field access.
-                    if (!((MetadataType)owningType).IsBeforeFieldInit)
+                    // For beforefieldinit, we can wait for field access...
+                    bool runCctor = !((MetadataType)owningType).IsBeforeFieldInit;
+
+                    // ...unless this is a PInvoke, in which case we run it regardless
```

Comment:
The logic that alters the behavior for optimizations on/off is not in the JIT. It is on the VM side: https://github.com/dotnet/coreclr/blob/7590378d8a00d7c29ade23fada2ce79f4495b889/src/vm/method.cpp#L2471 . I do not think that the JIT has any built-in policy on whether to trigger the cctor. It just does what the VM tells it to do.